File: apps/jvm/java-console/src/main/java/eu/sirotin/kotunil/app/java/Checker.java
package eu.sirotin.kotunil.app.java;

import eu.sirotin.kotunil.core.Expression;

import java.util.HashSet;
import java.util.Set;

public class Checker {

    private static final Set<String> objectTypes = new HashSet<>();
    private static final Set<Expression> objects = new HashSet<>();
    public static Integer numberTests = 0;

    public static Integer getNumberTestedObjects() {
        return objects.size();
    }

    public static Integer getNumberTestedObjectTypes() {
        return objectTypes.size();
    }

    public static<T1 extends Expression, T2 extends Expression> void check(T1 a, T2 b) {
        numberTests++;

        tryAddObject(a);
        tryAddObject(b);

        if (a.compareTo(b) == 0) return;
        throw new IllegalArgumentException("Expected '%s' but is '%s'".formatted(a, b));
    }

    private static <T1 extends Expression> void tryAddObject(T1 a) {
        String u = a.unitSymbols();
        objectTypes.add(u);
        objects.add(a);
    }

    public static void check(Double a, Double b) {
        double maxDif = 0.0000001;
        numberTests++;
        if (Math.abs(a - b) < maxDif) return;
        throw new IllegalArgumentException("Expected '%s' but is '%s'".formatted(a, b));
    }

    public static void check(String a, String b) {
        numberTests++;
        if (a.equals(b)) return;
        throw new IllegalArgumentException("Expected '%s' but is '%s'".formatted(a, b));
    }

    public static void check(Boolean a) {
        numberTests++;
        if (a) return;
        throw new IllegalArgumentException("Expected True but is False");
    }


}

File: apps/web_app_js/src/main/javascript/checker.js
var numberChecks = 0;
const EPS = 1e-10;
const objectTypes = new Set();
function checkValues(a, b){
    numberChecks++;
    if (typeof a === 'string'){
        if(a == b)return;
    }else{
        if(Math.abs(a - b) < EPS)return;
    }
    throw "Values are different a=" + a + " b=" + b;
}

function checkBoolean(a){
    numberChecks++;
    let t = typeof a;
    if(t === 'boolean'){
        if(a == true)return;
    }else {
        throw "Value not 'true'";
    }
    throw "Expected type 'boolean', but is: '" + t + "'";
}

function checkObjects(a, b){
    numberChecks++;
    let sa = a.unitSymbols();
    let sb = b.unitSymbols();
    objectTypes.add(sa);
    objectTypes.add(sb);
    if(sa != sb){
        throw "Objects have different types: '" + sa + "' vs '" + sb + "'";
    }
    if(Math.abs(a.value - b.value) < EPS){
        return;
    }
    throw "Values are different a=" + a + " b=" + b;
}

